operations.go

```diff
@@ -6,154 +6,160 @@ import (
 	"math/big"
 )
 
-type Int big.Int
+type Int struct {
+	Value *big.Int
+	Base  *big.Int
+}
 
 // Default prime (256-bit secp256k1 EC order)
 var (
-	p, _ = IntFromString("fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141", 16)
+	defaultP, _ = new(big.Int).SetString("fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141", 16)
 )
 
-// SetP resets the (prime) modulus across the package
-func SetP(new_prime *Int) *Int {
-	p = new_prime
-	return p
-}
-
-func GetP() *Int {
-	return p
-}
-
 // ARITHMETIC OPERATIONS
 
 // Add - Modular addition of (an arbitrary number of) field elements
-func (n *Int) Add(nums ...*Int) *Int {
-	out := big.NewInt(0)
-	for _, n := range nums {
-		out.Add(out, (*big.Int)(n))
-	}
-	out.Mod(out, (*big.Int)(p))
-	*n = (Int)(*out)
+func (n *Int) Add(x, y *Int) *Int {
+	v := new(big.Int).Add(x.Value, y.Value)
+	n.Value = v.Mod(v, x.Base)
+	n.Base = x.Base
```

Comment:
Here we are assigning pointers. We have to be confident if we are doing this. For instance, if we have two separate Ints in different threads and we change one of them, we are triggering a data race. I need to think it better, but at first glance it seems like it would be a better idea to hold copies.

Reply:
yes I thought about that. But the primes should stay static and never change so I though better not to have a million copies of the same prime. But I'm happy to do copies if necessary. The .Value, is always a copy (see how variable `v` was instantiated above). If you think the .Base should also be a copy I am happy to change that